docs/proposals/VIP-2.md

```diff
@@ -0,0 +1,240 @@
+>
+>Record 1: {_, 200, _}
+>
+>Record 2: {_, 201, _}
+>
+>Record 3: {5, _, _}
+>
+>Record 4: {_, _, 5001}
+
+Now, if we consume that heartbeat and read up to record 3, we'll have a high water mark of {5, 201, 4500}, and when looking at record 4, we can see that it has a vector component that is higher then our high water mark.
+
+The reason why the above works is because we're able to take for granted that we'll at some point either in the past or future get at least one update to help us build our high water mark.
+
+Now, the presented example is somewhat simplistic and there are some edge cases.  For example, what if our sequence of events was with a slightly different heartbeat. Something like:
+
+>HEARTBEAT: {_, 199, _}
+>
+>Record 1: {_, 200, _}
+>
+>Record 2: {_, 201, _}
+>
+>Record 3: {5, _, _}
+>
+>Record 4: {_, _, 5001}
+
+This is problematic. If reading from the heartbeat up to record 3, then we're stuck in the same situation as the minimal data example. So this informs that we need to be able to adhere to a property where when doing this kind of evaluation, we're able to assemble some information about all replicating colos. We can avoid this in all cases by making sure the heartbeat has all potentially revelevant information.  This would now look like:
+
+>HEARTBEAT: {3, 199, 4500}
+>
+>Record 1: {_, 200, _}
+>
+>Record 2: {_, 201, _}
+>
+>Record 3: {5, _, _}
+>
+>Record 4: {_, _, 5001}
+
+The heartbeat is now a published baseline that we can use to build our highwatermark when doing the comparison.
+
+**This approach is advantageous because it means we no longer have to merge together offset vectors in per row RMD's on venice servers. We need only have to publish on hand information to the version topic for a given store partiiton**
+
+#### Leap Frog with heartbeats (two colos)
+
+Lets look at what this implies for Leap Frog.  Leap Frog compares a high water mark and a record that both are remote from each other.
+
+>**COLO A**
+>
+>HEARTBEAT: {3, 199, 4500}
+>                           
+>Record 1: {_, 200, _}
+>
+>Record 2: {_, 201, _}
+>
+>Record 3: {5, _, _}
+>
+>Record 4: {_, _, 5001}
+
+>**COLO B**
+>
+>HEARTBEAT: {4, 199, 4900}
+>                           
+>Record 1: {_, 200, _}
+>
+>Record 2: {_, _, 4950}
+>
+>Record 3: {_, 300, _}
+>
+>Record 4: {20, _, _}
+
+Originally, we only needed the highwatermark of coloA and the rmd of an individual record in coloB (and vice versa going the other way). However, in this context, we need more information. Each record will only have the offset of the LAST upsream RT message which touched the row.  We can backfill a bit more information based on the last heartbeat. Since everything is produced in serial in increasing offsets we know that the other updates came from an offset previous to the published highwatermark in the last received heartbeat.  So, this looks like:
+
+>**COLO A**
+>
+>HEARTBEAT: {3, 199, 4500}
+>                           
+>Record 1: {<3, 200, <4500}
```

Comment:
For the backfill information like `<3`, does that mean we can always gurantee between HEARTBEAT and Record 1, there is no update from other regions. Only region #2 with offset 200 mute that key?

Reply:
I'll include a case here for the two updates as it's an important one. We have two choices, we can either rely on the heartbeat + previous events to build a running highwatermark from which we can apply this backfill, or, we can compact key updates which occur between two heartbeats.  

My intuition is that key updates are the way to go because if we apply an increasingly growing highwatermark into the offsets of individual records, then we narrow the common window between two colos. I've got a spec half done which models this, I'll post back here once I've determined conclusively that this intuition is right.  But it's a good call out because with two to the same key within the heartbeat interval then backfilling with the less then up to the last heartbeat is an incorrect generalization, and will lead to false positives in some simple cases.